src/test/scala/com/softwaremill/realworld/utils/TestUtils.scala

```diff
@@ -60,39 +60,61 @@ object TestUtils:
     Map("Authorization" -> ("Token " + jwt))
   }
 
-  def withEmptyDb(): RIO[TestDbLayer, Any] = for {
-    migrator <- ZIO.service[DbMigrator]
-    _ <- migrator.migrate()
-    _ <- loadFixture("fixtures/articles/admin.sql")
+  private def loadFixture(fixturePath: String): RIO[DataSource & Scope, Unit] = for {
+    _ <- ZIO.scope
+    dataSource <- ZIO.service[DataSource]
+    fixture <- ZIO.fromAutoCloseable(ZIO.attemptBlocking(Source.fromResource(fixturePath)))
+    connection <- ZIO.fromAutoCloseable(ZIO.attempt(dataSource.getConnection))
+    statement <- ZIO.fromAutoCloseable(ZIO.attempt(connection.createStatement()))
+  } yield {
+    val queries = fixture.mkString
+      .split(";")
+      .map(_.strip())
+      .filter(_.nonEmpty)
+
+    queries.foreach(statement.execute)
+  }
+
+  private def clearDb: RIO[DbConfig, Unit] = for {
+    cfg <- ZIO.service[DbConfig]
+    _ <- ZIO.attemptBlocking(Files.deleteIfExists(Paths.get(cfg.dbPath)))
   } yield ()
 
-  def withFixture(fixturePath: String): RIO[TestDbLayer, Any] = for {
+  private val initializeDb: RIO[DbMigrator, Unit] = for {
     migrator <- ZIO.service[DbMigrator]
     _ <- migrator.migrate()
+  } yield ()
+
+  private val withEmptyDb: RIO[DataSource with Scope with DbMigrator, Unit] = for {
+    _ <- initializeDb
     _ <- loadFixture("fixtures/articles/admin.sql")
+  } yield ()
+
+  private def withFixture(fixturePath: String): RIO[DataSource with Scope with DbMigrator, Unit] = for {
+    _ <- withEmptyDb
     _ <- loadFixture(fixturePath)
   } yield ()
 
-  def clearDb: RIO[TestDbLayer, Any] = for {
-    cfg <- ZIO.service[DbConfig]
-  } yield {
-    Files.deleteIfExists(Paths.get(cfg.dbPath))
-  }
+  /*
+  Clean up logic is packaged into a ZLayer.
+  This ensures that *.sqlite file will always get deleted after the test.
+  Clean up layer must be placed before the layer that provides a DataSource.
+  Placing clearDbLayer before Db.dataSourceLive guarantees that clean up will happen after DataSource is closed.
+   */
+  private val clearDbLayer = {
```

Comment:
are you running these tests on windows?
on MAC/ubuntu there is no problem with deleting these temp files but on Windows indeed there was - 
I'm just curious if you are resolving the same problem I faced in the past before I switched to MAC os :)
in my case these files couldn't be deleted because they were locked by the process

Reply:
I was running on Ubuntu. It seems to me that a call to .provide() at the top level supplied all tests with their own sqlite replica. Annotations, on the other hand, seem to work differently than providing layers. For example when you annotate a suite, clean up is performed once for the entire suite. In the end, more sqlite instances were created than deleted.